### ISSUE TITLE
Any easy way to get posixlib & libc to share fooOps?

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2995]
I am trying to take advantage of the fertile ground opened up by PR #2988, extern definitions.
In particular, I am trying to implement the Open Group 2018 specification for "net/in.h" for
both posixlib & libc with minimal duplication.   

My "if.scala"s are modeled on the "errno.scala"s & "signal.scala"s of PR #2988.

"in.h" deals with enumerating network interfaces, but the principle I have run aground upon should 
apply in general. I think theidiom i describe below is going to crop up in a number of places as we try
 to rationalize posixlib & libc.

"in.scala" defines 
```
type if_nameindex = CStruct2[
    CUnsignedInt, // if_index                                                   
    CString // if_name                                                          
  ]
```

Both posixlib & libc  `if` object  work in private testing.

libc "if.scala" follows the usual practice and defines "ifOps" to allow working with 
named fields of instances of if_nameidex.

My posixlib test programs use the implicit when I include libc.ifOps explicitly.
The also work when I duplicate the code of libc.ifOps in posixlib.ifOps.

What I would like to do is avoid that duplication or the hard to explain
"take this from here but to work with it take that from there".
Something like 
```
// posixlib net/if.scala

object ifOps { 
    import libc.net.ifOps.ifnameIndexOps._                                                 
  }

``` 

The "trait" idiom does not work. The compiler tells me I can not have an implicit class inside another class (i.e. trait).

I have reached the limits of my Scala foo here and would appreciate any instruction which would
allow me to avoid code duplication.   

Thank you.

**Comment by armanbilge** [comment]
You can do something like this:

```scala
class FooOps(val foo: Foo) extends AnyVal {
  def bar: Bar
}

trait FooSyntax {
  implicit def fooSyntax(foo: Foo): FooOps =
    new FooOps(foo)
}

object libcOps extends FooSyntax

object posixOps extends FooSyntax
```

**Comment by LeeTibbert** [comment]
Thank you. I will give it a run.

**Comment by LeeTibbert** [comment]
Works like a charm!  

Enumerating the network interfaces on a system, using either posixlib or libc.

```
Interface index 1 has name: 'lo'
Interface index 2 has name: 'en0'
```
IndexToName and NameToIndex both work.

**Comment by ekrich** [comment]
I guess I'm confused. `in.h` is not part of libc or is that a `struct` that is in a lib C definition. I am not seeing ant "net" in the C standard - https://en.cppreference.com/w/c/header

**Comment by LeeTibbert** [comment]
Probably flying similar but different flags:  https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/net_if.h.html.
`if.h` is used for TCP/IP network interfaces: enumerate them, translate network names to index numbers, translate
interface indices to names, etc.

I also do not see that file in the reference you posted.   I think you/we are looking at " C standard library" which
I suspect is a subset of the full set defined in the C standard:
```
    C17 standard (ISO/IEC 9899:2018): 
        7.1.2 Standard headers (p: 131-132) 
```

The only access I have to that standard is thru the Open Group URL I posted.

I am trying to implement `net/if.h` in both posixlib & libc because I would like to 
use it in a javalib `NetInterface.scala` I am working on.  Having the libc variant
would allow me to avoid introducing the kind of posixlib dependency that we are
trying to purge from javalib.

**Comment by LeeTibbert** [comment]
Looking at the URL you posted, it pertains to ISO/IEC C.   I believe that can and does
differ in slight ways from IEEE C.  I do not have a copy of the ISO C standard so I can
not cite chapter & verse.

To my thinking, SN is likely to be far better off not spending time splitting IEEE/ISO hairs but in
defining a lib for operating system and or  implementation specific (think of gnuisms) files, where each defines the 
definition/flag it is following.   "This is the standard stuff, this is the standard it is following" & "this is 
non-standard stuff (think epoll and kexec) and this is the definition it is following."

**Comment by LeeTibbert** [comment]
Later:  This afternoon I have been working on using the new "extern trait" to harmonize
posixlib & libc signal.scala (silly me!).

I've read this line from the Open Group POSIX/IEEE_C spec many times before, but it seems especially relevant today:
  The ISO C standard only requires the signal names SIGABRT, SIGFPE, SIGILL, SIGINT, SIGSEGV, and SIGTERM to
  be  defined. 

That is a concrete example of IEEE_C and ISO C differing.

**Comment by ekrich** [comment]
I don't know what exactly you are talking about. libc should be standard C like C11 and should work on both Unix style OSes and Windows and virtually everything out there. The Java lib beyond stuff supported by standard C (which is not much) should split between using POSIX and Windows like it does now. On the UNIX side we should try and use POSIX and avoid GNUisms. In third party libraries, anything goes but not in core Scala Native.

**Comment by armanbilge** [comment]
> The Java lib beyond stuff supported by standard C (which is not much) should split between using POSIX and Windows like it does now.

The problem is that there are several JDK APIs we simply cannot implement without GNUisms or BSDisms; POSIX is not enough.

**Comment by LeeTibbert** [comment]
C11 is ISO/IEC 9899:2011.  https://port70.net/~nsz/c/c11/n1570.html is a public description of the standard.
It Appendex B appears to be the original document which the https://en.cppreference.com/w/c/header URL 
above summarizes.

As you noted, `net/if.h` is not described there.  So, does that mean:

1) I can provide only a  Open Group (POSIX) posixlib implementation and
    do not have to worry about layering on `clib`?

2) I am free to use the posixlib include in a putative javalib `NetworkInterface.scala`?
    I am implementing Unix-like only while trying to leave the door open for
    a Windows implementation by someone else with that skill set.

If both of those are true, it immediately makes my life _much_ easier.   
 
Even with those, I think the need for an `osLib` or "NonStandardLib" is going
to become more urgent.  Close to bare metal rarely gets standardized (insert 
here the obligatory xkcd "We have 12 standards, we need One Standard to rule them all! 
Now we have 13 standards" cartoon).

**Comment by ekrich** [comment]
> 1. I can provide only a Open Group (POSIX) posixlib implementation and
do not have to worry about layering on clib?

Yes, any POSIX is fair game on its own. No layering on clib is needed; only for C/POSIX shared definitions.

> 2. I am free to use the posixlib include in a putative javalib NetworkInterface.scala?
I am implementing Unix-like only while trying to leave the door open for
a Windows implementation by someone else with that skill set.

Yes, you can use that for sure. If `NetworkInterface` is an implementation that is not officially a part of the `javalib` then it would be nice if the file / package is defined maybe in the `scala.scalanative` namespace in the javalib.

As far as Windows support, you can have a shared C functions like this: https://github.com/scala-native/scala-native/blob/main/nativelib/src/main/resources/scala-native/time_millis.c

The other alternative is to follow the `if()` standard like this - https://github.com/scala-native/scala-native/blob/main/javalib/src/main/scala/java/nio/file/Files.scala#L204-L225

It really just depends on what makes the most sense in the particular situation. Hope this helps.

**Comment by LeeTibbert** [comment]
Yes that helps, a lot.  `NetworkInterface` and the support class it requires are both specified in Java 8 thru at least 18.
So they can be in their natural "java.net" package.

For those following along in the future:
   There is a third way of doing separate Unix and Windows support, as implemented in javalib `ProcessBuilder.scala.

The two problems of trying to provide `NetworkInterface` for Windows are:
1) I do not have a Windows build environment
2) I do not have the knowledge and experience to develop on Windows.

So a Unix-like only `NetworkInterface` is, IMO, better than none at all. Particularly
when I have a request from someone who would like to be using it Now.
Actually not now, but a month prior to today.

As usual. I will ensure that CI on Windows does not break. I will also try
to leave the code tidy enough that a Windows implementation can be
provided by someone motivated, knowledgeable, and experienced.

There is a third way of doing things, as implemented in javalib ProcessBuilder.scala.

**Comment by LeeTibbert** [comment]
Fixed in PR #3000     

Thank you @ekrich & @armanbilge for helping get this done.
I hope to also use it to correct a small display defect in Inet6Address.
